### PR TITLE
calib3d: fix the test fail on Calib3d_SolvePnP.accuracy

### LINKED ISSUE #9187: Calib3d_SolvePnP.accuracy is too unstable based on random seed
##### System information (version)
- OpenCV => master (7733654, 20f603a)
- Operating System / Platform => Windows 7 64 Bit Core i7 / Ubuntu 14.04
- Compiler => Visual Studio 2013 / GCC 4.8

##### Detailed description
- The test ```Calib3d_SolvePnP.accuracy``` for method ```SOLVEPNP_P3P``` seems too strict for some input data
- I found the error on Jetson TK1 
```
[----------] 3 tests from Calib3d_SolvePnP
[ RUN      ] Calib3d_SolvePnP.accuracy
mode: 0, method: 0 -> 98.4% (err < 8.30464e-06)
mode: 0, method: 1 -> 95.6% (err < 1.34671e-05)
mode: 0, method: 2 -> 5.3% (err < 0.00592602)
mode: 0, method: 3 -> 99.9% (err < 0.000835733)
mode: 0, method: 4 -> 100% (err < 1.90072e-05)
mode: 0, method: 5 -> 98.8% (err < 0.00191663)
mode: 1, method: 0 -> 98.4% (err < 5.48124e+14)
mode: 1, method: 1 -> 96.1% (err < 0.352514)
mode: 1, method: 2 -> 5.8% (err < 0.00651514)
mode: 1, method: 3 -> 99.8% (err < 0.441405)
mode: 1, method: 4 -> 99.5% (err < 1.46951)
mode: 1, method: 5 -> 99.1% (err < 0.000364948)
/home/ubuntu/opencv-fork/modules/ts/src/ts.cpp:541: Failure
Failed

	failure reason: Bad accuracy
	test case #-1
	seed: ffffffffffffffff
-----------------------------------
	LOG:
Invalid accuracy for method 2, failed 947 tests from 1000, maximum error equals 0.005926, distortion mode equals 0
Invalid accuracy for method 2, failed 942 tests from 1000, maximum error equals 0.006515, distortion mode equals 1

-----------------------------------
[  FAILED  ] Calib3d_SolvePnP.accuracy (6192 ms)
[ RUN      ] Calib3d_SolvePnP.double_support
[       OK ] Calib3d_SolvePnP.double_support (2 ms)
[ RUN      ] Calib3d_SolvePnP.translation
[       OK ] Calib3d_SolvePnP.translation (4 ms)
[----------] 3 tests from Calib3d_SolvePnP (6208 ms total)
```
- I thought this problem comes from difference of instruction set and/or compiler
- Though during the investigation I realized that this doesn't happen even on exactly same platform/OS/program
   - For instance, when I pass ```--gtest_filter=Calib3d_SolvePnP*``` to gtest program to skip all the other tests, the test actually passes.
```

[----------] 3 tests from Calib3d_SolvePnP
[ RUN      ] Calib3d_SolvePnP.accuracy
mode: 0, method: 0 -> 98.6% (err < 9.75961e-06)
mode: 0, method: 1 -> 96.7% (err < 1.63235e-05)
mode: 0, method: 2 -> 89.6% (err < 0.0124706)
mode: 0, method: 3 -> 99.9% (err < 0.000706929)
mode: 0, method: 4 -> 100% (err < 1.67126e-05)
mode: 0, method: 5 -> 98.5% (err < 0.00132396)
mode: 1, method: 0 -> 98.2% (err < 7779.09)
mode: 1, method: 1 -> 96.3% (err < 0.219446)
mode: 1, method: 2 -> 88.7% (err < 0.00896903)
mode: 1, method: 3 -> 100% (err < 1.72653e-05)
mode: 1, method: 4 -> 99.4% (err < 1.62516)
mode: 1, method: 5 -> 98.8% (err < 0.00115234)
[       OK ] Calib3d_SolvePnP.accuracy (6132 ms)
[ RUN      ] Calib3d_SolvePnP.double_support
[       OK ] Calib3d_SolvePnP.double_support (3 ms)
[ RUN      ] Calib3d_SolvePnP.translation
[       OK ] Calib3d_SolvePnP.translation (4 ms)
[----------] 3 tests from Calib3d_SolvePnP (6148 ms total)
```
- I modified a bit and figured out that the error is based on input points, not the platforms
- In another word, it could happen on other platforms based on the random seed
- IMHO, [this threshold ](https://github.com/opencv/opencv/blob/773365405f146f15550ce65bcb1a0bd58d1fc163/modules/calib3d/test/test_solvepnp_ransac.cpp#L204) should be loosen to something like ```3.0e-3```

###### Visual Studio 2013
 - I pushed a branch in my repository ( f76e226ed6e7bff4d5809b03c787db8f35d701b7 ) and I added a test ```Calib3d_SolvePnP.verifyBehaviour``` in ```opencv_test_calib3d```
 - I hard coded a value which reproduces ```bad accuracy``` failure, and the value is in the range of generated random data

###### gcc (ARM)
 - I confirmed on following platforms
   - Jetson TK1, Jetson TX1, Jetson TX2, ODROID-C2, ODROID-X2, PINE64, 
   - As described above, it's not the problem of platform, it's the problem of run time.
   - For example, I used following cmake option for Jetson TK1
- ```cmake option``` : ```-DCPU_BASELINE=FP16 -DNEON=ON```
- the execution order of the test is
   - cameraCalibrationTiltTest
   - Calib3d_UndistortPoints
   - Calib3d_InitUndistortRectifyMap
   - Calib3d_Undistort
   - Calib3d_StereoBM
   - Calib3d_StereoSGBM
   - Calib3d_StereoSGBM_HH4
   - Calib3d_POSIT
   - Calib3d_CalibrateCamera_CPP
   - Calib3d_SolveP3P
   - Calib3d_SolvePnPRansac
   - Calib3d_SolvePnP

### PR DESCRIPTION
**Pull request by tomoaki0705**
### This pullrequest changes
resolves #9187

### CODE REVIEW

File: modules/calib3d/include/opencv2/calib3d.hpp [hunk 1]
```diff
@@ -237,7 +237,8 @@ enum { SOLVEPNP_ITERATIVE = 0,
        SOLVEPNP_P3P       = 2, //!< Complete Solution Classification for the Perspective-Three-Point Problem @cite gao2003complete
        SOLVEPNP_DLS       = 3, //!< A Direct Least-Squares (DLS) Method for PnP  @cite hesch2011direct
        SOLVEPNP_UPNP      = 4, //!< Exhaustive Linearization for Robust Camera Pose and Focal Length Estimation @cite penate2013exhaustive
-       SOLVEPNP_AP3P      = 5  //!< An Efficient Algebraic Solution to the Perspective-Three-Point Problem @cite Ke17
+       SOLVEPNP_AP3P      = 5, //!< An Efficient Algebraic Solution to the Perspective-Three-Point Problem @cite Ke17
+       SOLVEPNP_MAX_COUNT = 6, //!< Used for count
```
Comment (alalek) [inline modules/calib3d/include/opencv2/calib3d.hpp]: Please remove "value" and "comma":
```
    __SOLVEPNP_MAX_COUNT   //!< Used for count
```

File: modules/calib3d/test/test_solvepnp_ransac.cpp [hunk 2]
```diff
@@ -61,22 +61,22 @@ class CV_solvePnPRansac_Test : public cvtest::BaseTest
...
     ~CV_solvePnPRansac_Test() {}
 protected:
-    void generate3DPointCloud(vector<Point3f>& points, Point3f pmin = Point3f(-1,
-        -1, 5), Point3f pmax = Point3f(1, 1, 10))
+    void generate3DPointCloud(vector<Point3f>& points,
+        Point3f pmin = Point3f(-10, -10, 5),
+        Point3f pmax = Point3f(10, 10, 10))
     {
-        const Point3f delta = pmax - pmin;
+        RNG rng(0x99999999); // fix the seed to use "fixed" input 3D points
```
Comment (alalek) [inline modules/calib3d/test/test_solvepnp_ransac.cpp]: I believe we should not fix point cloud between different test cases. Try to use `cv::theRNG()` or `ts->get_rng()` instead.

File: modules/calib3d/test/test_solvepnp_ransac.cpp [hunk 3]
```diff
@@ -216,38 +215,49 @@ class CV_solvePnP_Test : public CV_solvePnPRansac_Test
...
-        }
-        else if(method == 3)
+        switch(method)
         {
-            opoints = std::vector<Point3f>(points.begin(), points.begin()+50);
+            case SOLVEPNP_P3P:
+            case SOLVEPNP_UPNP:
+                opoints = std::vector<Point3f>(points.begin(), points.begin()+4);
+                break;
+            case SOLVEPNP_AP3P:
```
Comment (alalek) [inline modules/calib3d/test/test_solvepnp_ransac.cpp]: SOLVEPNP_AP3P is swapped with SOLVEPNP_UPNP in this switch statement.

File: modules/calib3d/include/opencv2/calib3d.hpp (hunk 1 above)
Comment (tomoaki0705) [inline modules/calib3d/include/opencv2/calib3d.hpp]: Thank you. Fixed.

File: modules/calib3d/test/test_solvepnp_ransac.cpp (hunk 2 above)
Comment (tomoaki0705) [inline modules/calib3d/test/test_solvepnp_ransac.cpp]: Thank you. Fixed

File: modules/calib3d/test/test_solvepnp_ransac.cpp (hunk 3 above)
Comment (tomoaki0705) [inline modules/calib3d/test/test_solvepnp_ransac.cpp]: I made a mistake. Thank you for reviewing.

**Comment by tomoaki0705** [comment]
On my local, the ```mode == 0``` (no distortion) now passes.
But ```mode == 1``` (w/ distortion) fails under ```method == 2```
give me more time to investigate

**Comment by tomoaki0705** [comment]
One more thing.
The test passes if I loosen the ```eps``` of ```SOLVEPNP_P3P``` from ```1.0e-4``` to ```2.0e-4``` (just twice larger)

I think it's reasonable fix, but any opinion ?

**Comment by alalek** [comment]
I believe that eps increase is fine.

**Comment by tomoaki0705** [comment]
I really appreciate your nice review.
I squashed the commits to one.  If the build passes, this PR is ready to merge.

**Comment by alalek** [comment]
Thank you! :+1:

**Pull request merged**